### PR TITLE
heal: Send whitespaces to keep connections open

### LINKED ISSUE #4132: Trying to heal a disk with distributed Erasure Code
Hello, I wanted to help out and test the heal command on my production setup. I have a 6 node Raspberry Pi Cluster. Minio is running in a docker container on each host with 1 disk per node. 

I compiled the latest MC and MINIO server today from Git.

I started with 6 nodes then took two nodes offline and copied a lot of data to the nodes.  The cluster is working fine with 4/6 disks. Today. I added one of the downed nodes back to the cluster with a larger disk and tried to heal the disk and it fails

```
/mc admin heal myminio --json --debug

mc: <DEBUG> POST /?heal= HTTP/1.1
Host: 169.254.206.217:8001
User-Agent: Minio (linux; arm) madmin-go/0.0.1 mc/DEVELOPMENT.GOGET
Content-Length: 0
Authorization: AWS4-HMAC-SHA256 Credential=**REDACTED**/20170415/us-east-1/s3/aws4_request, SignedHeaders=host;x-amz-content-sha256;x-amz-date;x-minio-operation, Signature=**REDACTED**
X-Amz-Content-Sha256: e3b0c44298fc1c149afbf4c8996fb92427ae41e4649b934ca495991b7852b855
X-Amz-Date: 20170415T032758Z
X-Minio-Operation: format
Accept-Encoding: gzip

mc: <DEBUG> HTTP/1.1 500 Internal Server Error
Transfer-Encoding: chunked
Accept-Ranges: bytes
Content-Type: application/xml
Date: Sat, 15 Apr 2017 03:28:01 GMT
Server: Minio/DEVELOPMENT.GOGET (linux; arm)
Vary: Origin
X-Amz-Bucket-Region: us-east-1
X-Amz-Request-Id: 14B574713B72F657

102
<?xml version="1.0" encoding="UTF-8"?>
<Error><Code>InternalError</Code><Message>We encountered an internal error, please try again.</Message><Key></Key><BucketName></BucketName><Resource>/</Resource><RequestId>3L137</RequestId><HostId>3L137</HostId></Error>
0

mc: <DEBUG> Response Time:  3.007484862s

{"status":"error","error":{"message":"Cannot heal the disks.","cause":{"message":"We encountered an internal error, please try again.","error":{"Code":"InternalError","Message":"We encountered an internal error, please try again.","BucketName":"","Key":"","RequestID":"3L137","HostID":"3L137","Region":""}},"type":"fatal","trace":[{"line":408,"file":"admin-heal.go","func":"cmd.mainAdminHeal"}],"sysinfo":{"host.arch":"arm","host.cpus":"4","host.lang":"go1.7.5","host.name":"rpi-1","host.os":"linux","mem.heap.total":"2.7MB","mem.heap.used":"1.0MB","mem.total":"23MB","mem.used":"1.0MB"}}}
```

After adding the node back to the cluster, I still see the following on the nodes:
```
Drive Capacity: 26 GiB Free, 58 GiB Total
Status:         4 Online, 2 Offline. We can withstand [1] more drive failure(s).
```

Since I added the node back, it should say 5 online right? Does a heal have to happen first? Problem is that I can not heal =( I can access my files and write data when connecting to the web console on the added host.

### PR DESCRIPTION
**Pull request by vadmeste**
## Description
Heal objects can take time to finish, send whitespaces so clients
won't get timeout error.

## Motivation and Context
Fixes #4132 

## Types of changes

- [x] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)

## Checklist:

- [x] My code follows the code style of this project.
- [ ] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly.
- [x] I have added tests to cover my changes.
- [x] All new and existing tests passed.

### CODE REVIEW

File: cmd/admin-handlers.go [hunk 1]
```diff
@@ -425,6 +425,11 @@ func (adminAPI adminAPIHandlers) ClearLocksHandler(w http.ResponseWriter, r *htt
 	writeSuccessResponseJSON(w, jsonBytes)
 }
 
+type ListUploadsHealResponse struct {
```
Comment (harshavardhana) [inline cmd/admin-handlers.go]: Add a comment..

**Comment by krisis** [comment]
@vadmeste, the approach looks good to me. I shall review again once you include `newKeepConnUp` definition in the PR.

**Comment by vadmeste** [comment]
Ready for review.

File: cmd/admin-handlers.go [hunk 2]
```diff
@@ -778,25 +831,36 @@ func (adminAPI adminAPIHandlers) HealUploadHandler(w http.ResponseWriter, r *htt
...
-	if err != nil {
-		writeErrorResponse(w, toAPIErrorCode(err), r.URL)
-		return
+	} else {
+		jsonBytes, err := json.Marshal(newHealResult(numHealedDisks, numOfflineDisks))
+		if err != nil {
+			resp.Error = getAPIErrorFromCode(toAPIErrorCode(err), r.URL.Path)
+
+		} else {
+			resp.HealResult = jsonBytes
```
Comment (krisis) [inline cmd/admin-handlers.go]: IIUC, we are sending json value inside an xml response. Isn't that unconventional and possibly inconvenient for alternate minio admin SDK authors (in the future)?

File: cmd/api-errors.go [hunk 3]
```diff
@@ -717,3 +717,10 @@ func getAPIErrorResponse(err APIError, resource string) APIErrorResponse {
 		HostID:    "3L137",
 	}
 }
+
+// getAPIErrorFromCode converts API error code to ErrorResponse
+func getAPIErrorFromCode(errorCode APIErrorCode, resource string) *APIErrorResponse {
```
Comment (krisis) [inline cmd/api-errors.go:722]: `getAPIErrorResponse` would be better.

File: cmd/admin-handlers.go (hunk 2 above)
Comment (vadmeste) [reply cmd/admin-handlers.go]: > IIUC, we are sending json value inside an xml response. Isn't that unconventional and possibly inconvenient for alternate minio admin SDK authors (in the future)?

yes you are right, we probably need to change heal result from json to xml, what do you think ?

File: cmd/api-errors.go (hunk 3 above)
Comment (vadmeste) [reply cmd/api-errors.go:722]: > getAPIErrorResponse would be better.

We already have a function with the same exact name :) Looks above

Comment (krisis) [reply cmd/api-errors.go:722]: Oops :)

File: cmd/admin-handlers.go (hunk 2 above)
Comment (krisis) [reply cmd/admin-handlers.go]: As per the new admin REST API proposal [here](https://gist.github.com/krishnasrinivas/4b216a6ec03925645253788d8e0e241d), we are moving towards JSON encoding for all our responses. We should keep that in mind.

**Comment by deekoder** [comment]
Can this be merged. It's not clear to me whether the review is complete. Please help me understand.

**Comment by harshavardhana** [comment]
It is not ready to be merged yet @deekoder reviewing and testing.

**Comment by vadmeste** [comment]
I made some changes, no keep connection up structure anymore, just sending whitespaces the same way listening buckets sends notifications to clients.

File: cmd/admin-handlers.go [hunk 4]
```diff
@@ -53,6 +53,25 @@ const (
...
+// spaces at regular interval when no data are ready to be received.
+func sendXMLResponse(w http.ResponseWriter, data <-chan interface{}) {
+	writeSuccessResponseHeadersOnly(w)
+	for {
+		select {
+		case elem, ok := <-data:
+			if !ok {
+				return
+			}
+			writeBody(w, encodeResponse(elem))
```
Comment (harshavardhana) [inline cmd/admin-handlers.go]: We should also breakout of this loop if writes return error.

File: cmd/admin-handlers.go [hunk 5]
```diff
@@ -53,6 +53,25 @@ const (
...
+		case <-time.After(pingInterval):
+			writeBody(w, []byte(" "))
```
Comment (harshavardhana) [inline cmd/admin-handlers.go]: Breakout if the loop if the clients return error. They closed connection.

**Review by harshavardhana — CHANGES_REQUESTED** [review]

File: cmd/admin-handlers.go (hunk 5 above)
Comment (vadmeste) [reply cmd/admin-handlers.go]: Thanks, fixed now.

File: cmd/admin-handlers.go [hunk 6]
```diff
@@ -53,6 +53,29 @@ const (
 	mgmtUploadID       mgmtQueryKey = "upload-id"
 )
 
+const pingInterval = 5 * time.Second
```
Comment (harshavardhana) [inline cmd/admin-handlers.go]: Add a comment on why we choose this..

**Review by harshavardhana — APPROVED** [review]
Tested with large sets of directories, connections are properly open and wait..

**Comment by deekoder** [comment]
Need to know if this can be closed?

**Pull request closed without merging**